### PR TITLE
Promtail: improve behavior of partial lines

### PR DESCRIPTION
**Pull request by farodin91**
**What this PR does / why we need it**:

I started to see OOM after upgrading to a promtail version with #8497. So played around to make partial line behavior more stable. You can now truncate partial lines and configure max partial lines buffer.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Review by JStickler — APPROVED** [review]
[docs team] LGTM

**Comment by farodin91** [comment]
@JStickler  I did some refactoring.

**Comment by JStickler** [comment]
@farodin91 Docs still look fine, someone from engineering needs to review the code.

File: clients/pkg/logentry/stages/extensions.go [hunk 1]
```diff
@@ -101,8 +114,29 @@ func (c *cri) Run(entry chan Entry) chan Entry {
 	return in
 }
 
+func (c *cri) ensureTruncateIfRequired(e *Entry) {
+	if c.cfg.MaxPartialLineSizeTruncate && len(e.Line) > c.cfg.MaxPartialLineSize.Val() {
+		e.Line = e.Line[:c.cfg.MaxPartialLineSize.Val()]
+	}
+}
+
+// CriConfig contains the configuration for a dropStage
```
Comment (MichelHollands) [inline clients/pkg/logentry/stages/extensions.go:123]:
```suggestion
// CriConfig contains the configuration for the cri stage
```

File: docs/sources/clients/promtail/stages/cri.md [hunk 2]
```diff
@@ -9,7 +9,15 @@ The `cri` stage is a parsing stage that reads the log line using the standard CR
...
 ```yaml
-cri: {}
+cri:
+  # Max buffer size to hold partial lines.
+  [max_partial_lines: <int> | default = 100]
+
+  # Max line size to hold a single partial line, if max_partial_line_size_truncate is true. Example: 262144.
+  [max_partial_line_size: <int> | default = 0]
+
+  # Allows to pretruncate partial lines before stored in partial buffer.
```
Comment (MichelHollands) [inline docs/sources/clients/promtail/stages/cri.md:19]:
```suggestion
  # Allows to pretruncate partial lines before storing in partial buffer.
```

**Comment by MichelHollands** [comment]
@farodin91 Thanks for this contribution. I have a few comments about some wording. Can you fix the CI as indicated in the error message? Once that is done I'll merge this.

**Comment by farodin91** [comment]
@MichelHollands  Done.

File: clients/pkg/logentry/stages/extensions.go [hunk 3]
```diff
@@ -63,25 +68,30 @@ func (c *cri) Run(entry chan Entry) chan Entry {
...
-				e.Line = strings.Join([]string{prev.Line, e.Line}, "")
+				var builder strings.Builder
+				builder.WriteString(prev.Line)
+				builder.WriteString(e.Line)
+				e.Line = builder.String()
 			}
+			c.ensureTruncateIfRequired(&e)
 			c.partialLines[fingerprint] = e
 
 			return []Entry{e}, true // it's a partial-line so skip it.
```
Comment (Mehmoodhassan) [inline clients/pkg/logentry/stages/extensions.go:97]: Do promtail ignore all P(artial) log entries by default?

**Pull request merged**